### PR TITLE
chore: refactor customJs lib import flow for pg transaction support

### PR DESCRIPTION
**Pull request by AnaghHegde**
## Description
Please refer this document for more details - https://www.notion.so/appsmith/Transaction-Handling-in-PG-468cf8d4255749c3915699e59e91dc2f

## Automation

/ok-to-test tags="@tag.Git, @tag.ImportExport, @tag.Templates"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/9855863943>
> Commit: b9dfe1d758410e831f751f1fe1047373c1137513
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=9855863943&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Git, @tag.ImportExport, @tag.Templates`
> Spec:
> <hr>Tue, 09 Jul 2024 11:25:30 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
  - Introduced support for managing custom JavaScript libraries, including dry run operations.
- **Enhancements**
  - Enhanced dry operation handling with new methods and improved logic for managing custom JavaScript libraries.

### CODE REVIEW

**Review by abhvsn — COMMENTED** [review]
@AnaghHegde Changes looks fine but do we want to continue with separate maps for each entity? If not let's try and move away from it right now otherwise we are taking on debt early on in this refactor 🙏🏽.

**Comment by AnaghHegde** [comment]
@abhvsn I will work on this once i get more clarity on the page and actions refactor. These flows might require some changes hence i am going to stick to this approach for now.

**Review by abhvsn — APPROVED** [review]
Approving now to unblock, but let's address the debt while refactoring for pages and application import flow.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java [hunk 1]
```diff
@@ -73,11 +75,27 @@ public Mono<Boolean> addJSLibsToContext(
...
+            Boolean isForceInstall,
+            Map<String, List<CustomJSLib>> customJSLibsDryOps,
+            boolean isDryOps) {
         return repository
                 .findUniqueCustomJsLib(jsLib)
                 // Read more why Mono.defer is used here.
                 // https://stackoverflow.com/questions/54373920/mono-switchifempty-is-always-called
-                .switchIfEmpty(Mono.defer(() -> repository.save(jsLib)))
+                .switchIfEmpty(Mono.defer(() -> {
+                    if (isDryOps) {
```
Comment (sondermanish) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java:92]: QQ: could we think of a condition where the ops is dryOps but we expect the customJs to be present over there.
Maybe for a git app or something or importing something on existing app?

Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java:92]: If that is not the case then dryOps conditional could be taken outside. Although this way it handles more scenarios. However it's a bit slow

File: app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java [hunk 2]
```diff
@@ -86,6 +104,10 @@ public Mono<CustomJSLibContextDTO> persistCustomJSLibMetaDataIfDoesNotExistAndGe
                     */
                     if ((jsLib.getDefs().length() > foundJSLib.getDefs().length()) || isForceInstall) {
                         jsLib.setId(foundJSLib.getId());
+                        if (isDryOps) {
+                            addDryOpsForEntity("SAVE", customJSLibsDryOps, jsLib);
```
Comment (sondermanish) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java:108]: let's create an Enum to tackle the dryops scenario? if it's big of a task let's take it in separate PR.

Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java:108]: The ENUM is already present, i have added this in datasource PR. I will pull this changes and update it accordingly. Thanks for catching this ❤️

File: app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java (hunk 1 above)
Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java:92]: I am not sure if i follow this. But either way the switchIfEmpty condition handles this right?

Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java:93]: I understand that but i am trying to refactor with as less changes to business logic as possible. I know that we are taking debt here, but we will address this in the future.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java [hunk 3]
```diff
@@ -73,11 +76,27 @@ public Mono<Boolean> addJSLibsToContext(
...
+            boolean isDryOps) {
         return repository
                 .findUniqueCustomJsLib(jsLib)
                 // Read more why Mono.defer is used here.
                 // https://stackoverflow.com/questions/54373920/mono-switchifempty-is-always-called
-                .switchIfEmpty(Mono.defer(() -> repository.save(jsLib)))
+                .switchIfEmpty(Mono.defer(() -> {
+                    if (isDryOps) {
+                        jsLib.updateForBulkWriteOperation();
+                        addDryOpsForEntity("SAVE", customJSLibsDryOps, jsLib);
```
Comment (sondermanish) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/jslibs/base/CustomJSLibServiceCEImpl.java:95]: Enum over here as well

**Comment by sondermanish** [comment]
We are deferring unit tests to another PR

**Review by sondermanish — APPROVED** [review]

**Pull request merged**